File: go/core/indicators/indicators.go
// Package indicators computes the technical series the strategies share.
//
// It replaces four separate implementations of the same handful of functions —
// ATR alone existed in four projects, two of them subtly different in how they
// seeded the first value.
//
// # Representation
//
// Every function returns a []float64 the same length as its input, with
// math.NaN() in the positions before the indicator has warmed up. NaN is used
// rather than zero because zero is a legitimate indicator value and a
// zero-filled warm-up silently biases any average taken over the series.
//
// Price-valued outputs (SMA, EMA, ATR, Donchian, VWAP, pivots) are in the same
// minor units as their money.Money input. float64 holds those integers exactly
// up to 2^53, which is far beyond any realistic price, so no precision is lost;
// use ToMoney to convert a value back.
package indicators

import (
	"math"

	"github.com/althk/tradekit/go/core/domain"
	"github.com/althk/tradekit/go/core/money"
)

// ToMoney converts an indicator value back to minor units, rounding half away
// from zero. A NaN input returns 0, which callers must therefore guard against
// with IsValid rather than relying on the zero.
func ToMoney(v float64) money.Money {
	if math.IsNaN(v) || math.IsInf(v, 0) {
		return 0
	}
	return money.Money(math.Round(v))
}

// IsValid reports whether an indicator value has warmed up.
func IsValid(v float64) bool { return !math.IsNaN(v) }

// Closes extracts the closing prices of a candle series as indicator input.
func Closes(cs []domain.Candle) []float64 {
	return field(cs, func(c domain.Candle) money.Money { return c.Close })
}

// Highs extracts the high prices of a candle series.
func Highs(cs []domain.Candle) []float64 {
	return field(cs, func(c domain.Candle) money.Money { return c.High })
}

// Lows extracts the low prices of a candle series.
func Lows(cs []domain.Candle) []float64 {
	return field(cs, func(c domain.Candle) money.Money { return c.Low })
}

func field(cs []domain.Candle, get func(domain.Candle) money.Money) []float64 {
	out := make([]float64, len(cs))
	for i, c := range cs {
		out[i] = float64(get(c))
	}
	return out
}

func nanSlice(n int) []float64 {
	out := make([]float64, n)
	for i := range out {
		out[i] = math.NaN()
	}
	return out
}

// SMA is the simple moving average over the trailing period values.
func SMA(values []float64, period int) []float64 {
	out := nanSlice(len(values))
	if period <= 0 || len(values) < period {
		return out
	}
	var sum float64
	for i, v := range values {
		sum += v
		if i >= period {
			sum -= values[i-period]
		}
		if i >= period-1 {
			out[i] = sum / float64(period)
		}
	}
	return out
}

// EMA is the exponentially weighted moving average, seeded with the simple
// average of the first period values.
//
// Seeding matters and is the difference between two implementations this
// replaces: seeding with the first value instead makes the early series
// depend on how much history the caller happened to load.
func EMA(values []float64, period int) []float64 {
	out := nanSlice(len(values))
	if period <= 0 || len(values) < period {
		return out
	}
	k := 2.0 / float64(period+1)

	var seed float64
	for i := 0; i < period; i++ {
		seed += values[i]
	}
	prev := seed / float64(period)
	out[period-1] = prev

	for i := period; i < len(values); i++ {
		prev = (values[i]-prev)*k + prev
		out[i] = prev
	}
	return out
}

// TrueRange is the per-bar true range: the greatest of the bar's own span, the
// gap up from the previous close, and the gap down to it. The first bar has no
// previous close and is therefore NaN, not its own high-low span.
func TrueRange(cs []domain.Candle) []float64 {
	out := nanSlice(len(cs))
	for i := 1; i < len(cs); i++ {
		prevClose := float64(cs[i-1].Close)
		hl := float64(cs[i].High - cs[i].Low)
		hc := math.Abs(float64(cs[i].High) - prevClose)
		lc := math.Abs(float64(cs[i].Low) - prevClose)
		out[i] = math.Max(hl, math.Max(hc, lc))
	}
	return out
}

// ATR is Wilder's average true range: a simple average of the first period
// true ranges, then Wilder smoothing.
//
// Wilder smoothing, not an EMA with the same period, is what "ATR(14)" means
// everywhere it is quoted. Substituting an EMA gives a series roughly 7%
// tighter, which silently moves every ATR-derived stop.
func ATR(cs []domain.Candle, period int) []float64 {
	out := nanSlice(len(cs))
	tr := TrueRange(cs)
	// tr[0] is NaN, so the first period true ranges are tr[1..period].
	if period <= 0 || len(cs) < period+1 {
		return out
	}

	var seed float64
	for i := 1; i <= period; i++ {
		seed += tr[i]
	}
	prev := seed / float64(period)
	out[period] = prev

	for i := period + 1; i < len(cs); i++ {
		prev = (prev*float64(period-1) + tr[i]) / float64(period)
		out[i] = prev
	}
	return out
}

// RSI is Wilder's relative strength index, on a 0-100 scale.
func RSI(values []float64, period int) []float64 {
	out := nanSlice(len(values))
	if period <= 0 || len(values) < period+1 {
		return out
	}

	var gain, loss float64
	for i := 1; i <= period; i++ {
		switch d := values[i] - values[i-1]; {
		case d > 0:
			gain += d
		default:
			loss -= d
		}
	}
	avgGain, avgLoss := gain/float64(period), loss/float64(period)
	out[period] = rsiFrom(avgGain, avgLoss)

	for i := period + 1; i < len(values); i++ {
		d := values[i] - values[i-1]
		var g, l float64
		if d > 0 {
			g = d
		} else {
			l = -d
		}
		avgGain = (avgGain*float64(period-1) + g) / float64(period)
		avgLoss = (avgLoss*float64(period-1) + l) / float64(period)
		out[i] = rsiFrom(avgGain, avgLoss)
	}
	return out
}

func rsiFrom(avgGain, avgLoss float64) float64 {
	if avgLoss == 0 {
		if avgGain == 0 {
			return 50 // flat series: neither overbought nor oversold
		}
		return 100
	}
	rs := avgGain / avgLoss
	return 100 - 100/(1+rs)
}

// Channel is a Donchian channel: the highest high and lowest low over a
// trailing window, and the midpoint between them.
type Channel struct {
	Upper  []float64
	Lower  []float64
	Middle []float64
}

// Donchian computes the channel over the trailing period bars.
//
// The window excludes the current bar. A breakout strategy that includes it can
// never fire: the current bar's own high is by definition not greater than the
// highest high of a window containing it.
func Donchian(cs []domain.Candle, period int) Channel {
	ch := Channel{Upper: nanSlice(len(cs)), Lower: nanSlice(len(cs)), Middle: nanSlice(len(cs))}
	if period <= 0 || len(cs) <= period {
		return ch
	}
	for i := period; i < len(cs); i++ {
		hi, lo := float64(cs[i-period].High), float64(cs[i-period].Low)
		for j := i - period + 1; j < i; j++ {
			hi = math.Max(hi, float64(cs[j].High))
			lo = math.Min(lo, float64(cs[j].Low))
		}
		ch.Upper[i], ch.Lower[i] = hi, lo
		ch.Middle[i] = (hi + lo) / 2
	}
	return ch
}

// VWAP is the volume-weighted average price, accumulated from the start of
// each session.
//
// Sessions are delimited by newSession, which reports whether the bar at index
// i opens a new session. Passing a function rather than a calendar keeps this
// package free of a calendar dependency and lets a backtest group bars however
// its data is shaped.
func VWAP(cs []domain.Candle, newSession func(i int) bool) []float64 {
	out := nanSlice(len(cs))
	var pv, vol float64
	for i, c := range cs {
		if i == 0 || newSession(i) {
			pv, vol = 0, 0
		}
		typical := float64(c.High+c.Low+c.Close) / 3
		v := float64(c.Volume)
		pv += typical * v
		vol += v
		if vol > 0 {
			out[i] = pv / vol
		} else {
			// A zero-volume session has no volume-weighted price. Fall
			// back to the typical price rather than dividing by zero,
			// which is what an illiquid first bar would otherwise do.
			out[i] = typical
		}
	}
	return out
}

// Pivots is the central pivot range and the classic support and resistance
// levels derived from one completed session.
type Pivots struct {
	Pivot      money.Money
	BC, TC     money.Money // central pivot range bottom and top
	R1, R2, R3 money.Money
	S1, S2, S3 money.Money
}

// CPR computes the pivot levels for the session following the given candle.
//
// BC and TC are ordered so that BC is always the lower of the two; the raw
// formulas produce them in either order depending on where the close sat, and
// a caller comparing price to "the top of the range" needs them sorted.
func CPR(prev domain.Candle) Pivots {
	h, l, c := float64(prev.High), float64(prev.Low), float64(prev.Close)
	pivot := (h + l + c) / 3
	bc := (h + l) / 2
	tc := pivot - bc + pivot
	if bc > tc {
		bc, tc = tc, bc
	}
	return Pivots{
		Pivot: ToMoney(pivot),
		BC:    ToMoney(bc),
		TC:    ToMoney(tc),
		R1:    ToMoney(2*pivot - l),
		S1:    ToMoney(2*pivot - h),
		R2:    ToMoney(pivot + (h - l)),
		S2:    ToMoney(pivot - (h - l)),
		R3:    ToMoney(h + 2*(pivot-l)),
		S3:    ToMoney(l - 2*(h-pivot)),
	}
}

// ADX is Wilder's average directional index together with the directional
// indicators it is built from, all on a 0-100 scale.
type ADX struct {
	ADX     []float64
	PlusDI  []float64
	MinusDI []float64
}

// ComputeADX returns the ADX and directional indicators over the given period.
func ComputeADX(cs []domain.Candle, period int) ADX {
	n := len(cs)
	res := ADX{ADX: nanSlice(n), PlusDI: nanSlice(n), MinusDI: nanSlice(n)}
	if period <= 0 || n < 2*period+1 {
		return res
	}

	tr := TrueRange(cs)
	plusDM, minusDM := make([]float64, n), make([]float64, n)
	for i := 1; i < n; i++ {
		up := float64(cs[i].High - cs[i-1].High)
		down := float64(cs[i-1].Low - cs[i].Low)
		if up > down && up > 0 {
			plusDM[i] = up
		}
		if down > up && down > 0 {
			minusDM[i] = down
		}
	}

	var sTR, sPlus, sMinus float64
	for i := 1; i <= period; i++ {
		sTR += tr[i]
		sPlus += plusDM[i]
		sMinus += minusDM[i]
	}

	dx := nanSlice(n)
	setDI := func(i int) {
		if sTR == 0 {
			res.PlusDI[i], res.MinusDI[i], dx[i] = 0, 0, 0
			return
		}
		p := 100 * sPlus / sTR
		m := 100 * sMinus / sTR
		res.PlusDI[i], res.MinusDI[i] = p, m
		if p+m == 0 {
			dx[i] = 0
		} else {
			dx[i] = 100 * math.Abs(p-m) / (p + m)
		}
	}
	setDI(period)

	for i := period + 1; i < n; i++ {
		sTR = sTR - sTR/float64(period) + tr[i]
		sPlus = sPlus - sPlus/float64(period) + plusDM[i]
		sMinus = sMinus - sMinus/float64(period) + minusDM[i]
		setDI(i)
	}

	// ADX is Wilder's average of DX, first available a further period on.
	first := 2 * period
	var seed float64
	for i := period; i <= first; i++ {
		seed += dx[i]
	}
	prev := seed / float64(period+1)
	res.ADX[first] = prev
	for i := first + 1; i < n; i++ {
		prev = (prev*float64(period-1) + dx[i]) / float64(period)
		res.ADX[i] = prev
	}
	return res
}

File: py/src/tradekit/core/indicators.py
"""Technical series shared by the strategies.

Mirrors ``go/core/indicators``. It replaces four separate implementations of the
same handful of functions -- ATR alone existed in four projects, two of them
subtly different in how they seeded the first value.

Representation
--------------
Every function returns a list of floats the same length as its input, with
``float("nan")`` in the positions before the indicator has warmed up. NaN is
used rather than zero because zero is a legitimate indicator value, and a
zero-filled warm-up silently biases any average taken over the series.

Price-valued outputs (SMA, EMA, ATR, Donchian, VWAP, pivots) are in the same
minor units as their :class:`~tradekit.core.money.Money` input. A float holds
those integers exactly up to 2**53, far beyond any realistic price, so no
precision is lost; use :func:`to_money` to convert a value back.
"""

from __future__ import annotations

import math
from collections.abc import Callable, Sequence
from dataclasses import dataclass

from .domain import Candle
from .money import Money, _round_half_away

__all__ = [
    "ADX",
    "Channel",
    "Pivots",
    "adx",
    "atr",
    "closes",
    "cpr",
    "donchian",
    "ema",
    "highs",
    "is_valid",
    "lows",
    "rsi",
    "sma",
    "to_money",
    "true_range",
    "vwap",
]

NAN = float("nan")


def to_money(v: float) -> Money:
    """Convert an indicator value to minor units, rounding half away from zero.

    A NaN input returns zero, so callers must guard with :func:`is_valid` rather
    than relying on the zero to mean anything.
    """
    if math.isnan(v) or math.isinf(v):
        return Money(0)
    return Money(_round_half_away(v))


def is_valid(v: float) -> bool:
    """Whether an indicator value has warmed up."""
    return not math.isnan(v)


def closes(candles: Sequence[Candle]) -> list[float]:
    """Closing prices as indicator input."""
    return [float(c.close) for c in candles]


def highs(candles: Sequence[Candle]) -> list[float]:
    """High prices as indicator input."""
    return [float(c.high) for c in candles]


def lows(candles: Sequence[Candle]) -> list[float]:
    """Low prices as indicator input."""
    return [float(c.low) for c in candles]


def sma(values: Sequence[float], period: int) -> list[float]:
    """Simple moving average over the trailing ``period`` values."""
    out = [NAN] * len(values)
    if period <= 0 or len(values) < period:
        return out
    total = 0.0
    for i, v in enumerate(values):
        total += v
        if i >= period:
            total -= values[i - period]
        if i >= period - 1:
            out[i] = total / period
    return out


def ema(values: Sequence[float], period: int) -> list[float]:
    """Exponential moving average, seeded with the mean of the first window.

    Seeding matters, and is the difference between two implementations this
    replaces: seeding with the first value instead makes the early series depend
    on how much history the caller happened to load.
    """
    out = [NAN] * len(values)
    if period <= 0 or len(values) < period:
        return out
    k = 2.0 / (period + 1)
    prev = sum(values[:period]) / period
    out[period - 1] = prev
    for i in range(period, len(values)):
        prev = (values[i] - prev) * k + prev
        out[i] = prev
    return out


def true_range(candles: Sequence[Candle]) -> list[float]:
    """Per-bar true range.

    The greatest of the bar's own span, the gap up from the previous close, and
    the gap down to it. The first bar has no previous close and is therefore
    NaN, not its own high-low span.
    """
    out = [NAN] * len(candles)
    for i in range(1, len(candles)):
        prev_close = float(candles[i - 1].close)
        out[i] = max(
            float(candles[i].high - candles[i].low),
            abs(float(candles[i].high) - prev_close),
            abs(float(candles[i].low) - prev_close),
        )
    return out


def atr(candles: Sequence[Candle], period: int) -> list[float]:
    """Wilder's average true range.

    A simple average of the first ``period`` true ranges, then Wilder smoothing.
    Wilder smoothing -- not an EMA of the same period -- is what "ATR(14)" means
    everywhere it is quoted; substituting an EMA gives a tighter series, which
    silently moves every ATR-derived stop.
    """
    out = [NAN] * len(candles)
    tr = true_range(candles)
    if period <= 0 or len(candles) < period + 1:
        return out
    prev = sum(tr[1 : period + 1]) / period
    out[period] = prev
    for i in range(period + 1, len(candles)):
        prev = (prev * (period - 1) + tr[i]) / period
        out[i] = prev
    return out


def rsi(values: Sequence[float], period: int) -> list[float]:
    """Wilder's relative strength index, on a 0-100 scale."""
    out = [NAN] * len(values)
    if period <= 0 or len(values) < period + 1:
        return out

    gain = loss = 0.0
    for i in range(1, period + 1):
        delta = values[i] - values[i - 1]
        if delta > 0:
            gain += delta
        else:
            loss -= delta
    avg_gain, avg_loss = gain / period, loss / period
    out[period] = _rsi_from(avg_gain, avg_loss)

    for i in range(period + 1, len(values)):
        delta = values[i] - values[i - 1]
        g = delta if delta > 0 else 0.0
        loss_step = -delta if delta < 0 else 0.0
        avg_gain = (avg_gain * (period - 1) + g) / period
        avg_loss = (avg_loss * (period - 1) + loss_step) / period
        out[i] = _rsi_from(avg_gain, avg_loss)
    return out


def _rsi_from(avg_gain: float, avg_loss: float) -> float:
    if avg_loss == 0:
        return 50.0 if avg_gain == 0 else 100.0
    rs = avg_gain / avg_loss
    return 100 - 100 / (1 + rs)


@dataclass(slots=True)
class Channel:
    """A Donchian channel: trailing highest high, lowest low, and midpoint."""

    upper: list[float]
    lower: list[float]
    middle: list[float]


def donchian(candles: Sequence[Candle], period: int) -> Channel:
    """Donchian channel over the trailing ``period`` bars.

    The window excludes the current bar. A breakout strategy that includes it can
    never fire: the current bar's own high is by definition not greater than the
    highest high of a window containing it.
    """
    n = len(candles)
    ch = Channel([NAN] * n, [NAN] * n, [NAN] * n)
    if period <= 0 or n <= period:
        return ch
    for i in range(period, n):
        window = candles[i - period : i]
        hi = max(float(c.high) for c in window)
        lo = min(float(c.low) for c in window)
        ch.upper[i], ch.lower[i], ch.middle[i] = hi, lo, (hi + lo) / 2
    return ch


def vwap(candles: Sequence[Candle], new_session: Callable[[int], bool]) -> list[float]:
    """Volume-weighted average price, accumulated from each session's start.

    ``new_session(i)`` reports whether the bar at index ``i`` opens a new
    session. Passing a callable rather than a calendar keeps this module free of
    a calendar dependency and lets a backtest group bars however its data is
    shaped.
    """
    out = [NAN] * len(candles)
    pv = vol = 0.0
    for i, c in enumerate(candles):
        if i == 0 or new_session(i):
            pv = vol = 0.0
        typical = float(c.high + c.low + c.close) / 3
        v = float(c.volume)
        pv += typical * v
        vol += v
        # A zero-volume session has no volume-weighted price; the typical price
        # is the honest fallback, and avoids dividing by zero on an illiquid bar.
        out[i] = pv / vol if vol > 0 else typical
    return out


@dataclass(slots=True)
class Pivots:
    """Central pivot range and the classic support/resistance levels."""

    pivot: Money
    bc: Money
    tc: Money
    r1: Money
    r2: Money
    r3: Money
    s1: Money
    s2: Money
    s3: Money


def cpr(prev: Candle) -> Pivots:
    """Pivot levels for the session following ``prev``.

    ``bc`` and ``tc`` are ordered so ``bc`` is always the lower of the two. The
    raw formulas produce them in either order depending on where the close sat,
    and a caller comparing price to "the top of the range" needs them sorted.
    """
    h, low, c = float(prev.high), float(prev.low), float(prev.close)
    pivot = (h + low + c) / 3
    bc = (h + low) / 2
    tc = pivot - bc + pivot
    if bc > tc:
        bc, tc = tc, bc
    return Pivots(
        pivot=to_money(pivot),
        bc=to_money(bc),
        tc=to_money(tc),
        r1=to_money(2 * pivot - low),
        s1=to_money(2 * pivot - h),
        r2=to_money(pivot + (h - low)),
        s2=to_money(pivot - (h - low)),
        r3=to_money(h + 2 * (pivot - low)),
        s3=to_money(low - 2 * (h - pivot)),
    )


@dataclass(slots=True)
class ADX:
    """Average directional index with its directional indicators, 0-100."""

    adx: list[float]
    plus_di: list[float]
    minus_di: list[float]


def adx(candles: Sequence[Candle], period: int) -> ADX:
    """Wilder's ADX and the +DI / -DI series it is built from."""
    n = len(candles)
    res = ADX([NAN] * n, [NAN] * n, [NAN] * n)
    if period <= 0 or n < 2 * period + 1:
        return res

    tr = true_range(candles)
    plus_dm, minus_dm = [0.0] * n, [0.0] * n
    for i in range(1, n):
        up = float(candles[i].high - candles[i - 1].high)
        down = float(candles[i - 1].low - candles[i].low)
        if up > down and up > 0:
            plus_dm[i] = up
        if down > up and down > 0:
            minus_dm[i] = down

    s_tr = sum(tr[1 : period + 1])
    s_plus = sum(plus_dm[1 : period + 1])
    s_minus = sum(minus_dm[1 : period + 1])
    dx = [NAN] * n

    def set_di(i: int) -> None:
        nonlocal s_tr, s_plus, s_minus
        if s_tr == 0:
            res.plus_di[i] = res.minus_di[i] = dx[i] = 0.0
            return
        p = 100 * s_plus / s_tr
        m = 100 * s_minus / s_tr
        res.plus_di[i], res.minus_di[i] = p, m
        dx[i] = 0.0 if p + m == 0 else 100 * abs(p - m) / (p + m)

    set_di(period)
    for i in range(period + 1, n):
        s_tr = s_tr - s_tr / period + tr[i]
        s_plus = s_plus - s_plus / period + plus_dm[i]
        s_minus = s_minus - s_minus / period + minus_dm[i]
        set_di(i)

    first = 2 * period
    prev = sum(dx[period : first + 1]) / (period + 1)
    res.adx[first] = prev
    for i in range(first + 1, n):
        prev = (prev * (period - 1) + dx[i]) / period
        res.adx[i] = prev
    return res
